### PR TITLE
Deprecate env config option

### LINKED ISSUE #5276: [7.0] Deprecate env option in .babelrc
~~## EDIT: We're just going to deprecate now~~
## Fixing merge behavior instead

Original issue details: 
<details>
http://babeljs.io/docs/usage/babelrc/#env-option

some env issues: https://github.com/babel/babel/issues/5275, https://github.com/babel/babel/issues/4539, https://github.com/babel/babel/issues/4013, https://github.com/babel/babel/issues/4817

```js
{
  "env": {
    "production": {
      "plugins": ["transform-react-constant-elements"]
    }
  }
}
```

Options specific to a certain environment are merged into and overwrite non-env specific options.

The env key will be taken from process.env.BABEL_ENV, when this is not available then it uses process.env.NODE_ENV if even that is not available then it defaults to "development".

- [ ] remove the merge behavior (each env needs to specify every config value again)
- [ ] drop the env option entirely and recommend using `.babelrc.js` using a combiation of ternaries + process.env manually (https://github.com/babel/babel/pull/4892)
- [ ] only allow a development, production, profiling env options https://twitter.com/dan_abramov/status/828997631561039872
- [ ] remove development as the default environment https://github.com/babel/babel/blob/badce969dafaf3128104cbc6d6bc605e2b47a3f5/packages/babel-preset-react/src/index.js#L22
- [ ] better validation from babel's side if plugins/presets use env
- [ ] ?

> true that allowing .js may lead to crazy configs but think it's better than confusing with babelrc/env
</details>

Because it's JS you could do this in a lot of ways

```js
var env = process.env.BABEL_ENV || process.env.NODE_ENV;
```

```js
// inline plugin
module.exports = {
  plugins: [
    env === 'development' && "transform-react-constant-elements"
  ].filter(Boolean)
};
```

https://github.com/facebookincubator/create-react-app/blob/65e63403952f4f3c7e872f707fb3736e339254d9/packages/babel-preset-react-app/index.js#L42-L51
```js
var plugins = [];
if (env === 'development') {
  plugins.push.apply(plugins, ["transform-react-constant-elements"]);
}
module.exports = { plugins };
```

### PR DESCRIPTION
**Pull request by kaicataldo**
| Q                        | A 
| ------------------------ | ---
| Patch: Bug Fix?          | no
| Major: Breaking Change?  | no
| Minor: New Feature?      | yes
| Deprecations?            | yes
| Spec Compliancy?         | no
| Tests Added/Pass?        | yes
| Fixed Tickets            | Fixes #5276
| License                  | MIT
| Doc PR                   | 
| Dependency Changes       | 

This logs a warning that `env` is now deprecated and will be removed in a future version of Babel.

### CODE REVIEW

File: packages/babel-core/test/config-chain.js [hunk 1]
```diff
@@ -533,14 +536,22 @@ describe("buildConfigChain", function () {
...
       /Error while parsing JSON/
     );
   });
+
+  it("env-deprecated", function() {
+    buildConfigChain({
+      filename: fixture("env-deprecated", "src.js"),
+    });
+
+    assert(console.log.calledWithMatch(/^The "env" option has been deprecated/));
```
Comment (kaicataldo) [inline packages/babel-core/test/config-chain.js]: So, I'm a bit confused by this. When I stub `console.log` in this file it errors saying that it's already been stubbed. This also happens when I skip the other file that I stub `console.log`, so I think I'm cleaning up after myself correctly. Anyone else see/know what might be causing this?

File: packages/babel-core/src/transformation/file/options/build-config-chain.js [hunk 2]
```diff
@@ -199,6 +206,14 @@ class ConfigChainBuilder {
 
     const envKey = babel.getEnv();
     if (options.env) {
+      // TODO: Remove in Babel v8
+      if (!warnedDeprecatedEnv) {
+        console.log(`The "env" option has been deprecated and will be removed in a future version of Babel.
```
Comment (xtuc) [inline packages/babel-core/src/transformation/file/options/build-config-chain.js]: Could we use https://github.com/babel/babel/blob/97cd51995bfd999c82a4fb2da933668683bd43d2/packages/babel-core/src/transformation/file/logger.js#L41 there?

Comment (existentialism) [inline packages/babel-core/src/transformation/file/options/build-config-chain.js]: Logger class is [likely going away](https://github.com/babel/babel/pull/5467)

Comment (kaicataldo) [inline packages/babel-core/src/transformation/file/options/build-config-chain.js]: Ah, it loos like that linked PR also touches the `env` logic, so this PR might not be what we want anyway.

**Pull request closed without merging**